src/group_theory/finiteness.lean

```diff
@@ -101,6 +102,46 @@ add_monoid.fg_iff_mul_fg.1 ‹_›
 
 end monoid
 
+namespace monoid
+
+open submonoid
+
+@[simp, to_additive]
+lemma fg_iff_submonoid_fg (N : submonoid M) : fg N ↔ N.fg :=
+begin
+  classical,
+  split,
+  { rintro ⟨s, hs⟩,
+    refine ⟨s.image N.subtype, _⟩,
+    rw [finset.coe_image, ← monoid_hom.map_mclosure, hs,
+      ← monoid_hom.mrange_eq_map, range_subtype] },
```

Comment:
There's a helper lemma you can extract here,
```lean
@[to_additive]
lemma submonoid.fg.map {M' : Type*} [monoid M'] {P : submonoid M} (h : P.fg) (e : M →* M') :
  (P.map e).fg :=
begin
  classical,
  obtain ⟨s, rfl⟩ := h,
  refine ⟨s.image e, by rw [finset.coe_image, monoid_hom.map_mclosure]⟩,
end
```
which makes these lines
```suggestion
  { intro h,
    convert h.out.map N.subtype,
    rw [← monoid_hom.mrange_eq_map, submonoid.range_subtype] },
```

Reply:
Similarly, if you transfer `submodule.fg_of_fg_map_injective` to submonoid, then this provides the other half of the proof